File: 46-permutations/permutations.ts
function permute(nums: number[]): number[][] {
    const results: number[][] = []

    const backtrack = (currentPermutation: number[], used: boolean[]) => {
        if (currentPermutation.length === nums.length) {
            results.push([...currentPermutation])
            return
        }

        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue

            currentPermutation.push(nums[i])
            used[i] = true

            backtrack(currentPermutation, used)

            currentPermutation.pop()
            used[i] = false
        }
    }

    backtrack([], new Array(nums.length).fill(false))

    return results
};

File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums):
        

        def backtrack(nums, candidate, used, res):
            if len(candidate) == len(nums):
                res.append(candidate[:])
                return

            for num in nums:
                if num not in used:
                    candidate.append(num)
                    used.add(num)

                    backtrack(nums, candidate, used, res)

                    candidate.pop()
                    used.remove(num)

        res = []
        backtrack(nums, [], set(), res)
        return res